SE-0425 implementation

stephentyrone opened the pull request:
Implements Int128 per SE-0425, and adopts it internally in Duration.

stephentyrone commented:
@swift-ci test

stephentyrone commented on stdlib/public/core/Int128.swift:
```diff
@@ -0,0 +1,414 @@
+  
+  @inlinable
+  public init?<T>(exactly source: T) where T: BinaryInteger {
+    guard let high = Int64(exactly: source >> 64) else { return nil }
```
We get that from Int64(exactly:) here. There's no need to check twice.

stephentyrone replied:
(There's a little bit of a tradeoff of a potentially expensive bignum shift vs. conjuring min/max values to compare against, but for modest-sized arguments it doesn't matter much either way, and we can do further optimization work in the future.)

stephentyrone commented:
@swift-ci smoke test

stephentyrone commented:
@swift-ci smoke test

natecook1000 commented:
`Codable` changes LGTM!

stephentyrone commented:
@swift-ci test

benrimmington reviewed (commented):
The `mutating` modifier is missing, in two default implementations.

stephentyrone commented:
@swift-ci smoke test

oscbyspro commented on stdlib/public/core/UInt128.swift:
```diff
@@ -0,0 +1,411 @@
+    @inlinable
+    get {
+      precondition(position >= 0 && position < endIndex)
+      return withUnsafeBytes(of: _value) {
+#if _endian(little)
+        $0.assumingMemoryBound(to: UInt.self)[position]
+#else
+        $0.assumingMemoryBound(to: UInt.self)[count - 1 - position]
+#endif
+      }
```
I believe you must use load(fromByteOffset:as:) to obey the abstract machine's definition of memory bound to T, since the memory at the location will be bound to UInt128 rather than UInt.

stephentyrone replied:
Rolled into the latest update, thanks.

stephentyrone commented on stdlib/public/core/UInt128.swift:
```diff
@@ -0,0 +1,411 @@
+  @inlinable
+  public init<T>(_truncatingBits source: T) where T: BinaryInteger {
+    self.init(T(truncatingIfNeeded: source))
+  }
```
Rolled into the latest update, thanks.

stephentyrone commented on stdlib/public/core/Int128.swift:
```diff
@@ -0,0 +1,386 @@
+  @inlinable
+  public init<T>(_truncatingBits source: T) where T: BinaryInteger {
+    self.init(T(truncatingIfNeeded: source))
+  }
```
Rolled into the latest update, thanks.

stephentyrone commented on stdlib/public/core/Int128.swift:
```diff
@@ -0,0 +1,386 @@
+  
+  @inlinable
+  public var magnitude: Magnitude {
+    let unsignedSelf = UInt128(truncatingIfNeeded: self)
```
Rolled into the latest update, thanks.

stephentyrone commented:
@swift-ci smoke test

stephentyrone commented:
@swift-ci test

stephentyrone committed:
Implement Int128 and UInt128 per SE-0425

Also adopts these types in the implementation of Duration, and makes &* a customization point on FixedWidthInteger.

stephentyrone commented:
@swift-ci test

stephentyrone commented on stdlib/public/core/Int128.swift:
```diff
@@ -0,0 +1,386 @@
+  
+  public var byteSwapped: Self {
+    return Self(low: UInt64(truncatingIfNeeded: high.byteSwapped),
+                high: Int64(truncatingIfNeeded: low.byteSwapped))
```
This is resolved now.

stephentyrone commented on stdlib/public/core/Duration.swift:
```diff
@@ -77,12 +78,14 @@ public struct Duration: Sendable {
   ///   - attosecondsComponent: The attosecond component portion of the 
   ///                           `Duration` value.
   public init(secondsComponent: Int64, attosecondsComponent: Int64) {
+    guard #available(SwiftStdlib 6.0, *) else { fatalError() }
```
We'll keep _[U]Int128 around for the 6.0 implementation to mitigate this, but we really need to be able to move forward and not carry it forever.

stephentyrone commented on stdlib/public/core/Codable.swift:
```diff
@@ -3699,6 +4071,13 @@ internal final class _KeyedEncodingContainerBox<
+  
+  @available(SwiftStdlib 6.0, *)
+  override internal func encode<K: CodingKey>(_ value: Int128, forKey key: K) throws {
+    _internalInvariant(K.self == Key.self)
+    let key = unsafeBitCast(key, to: Key.self)
```
I think that I'm going to leave it as-is to keep this PR focused and not introduce differences from the surrounding code (but I think a follow-on PR that converts all of these is definitely appropriate!)

stephentyrone committed:
Revert adoption of Int128 in Duration for now

It would prevent using toolchain downloads on macOS 14 (which is
really a bug in how toolchains are configured that should be fixed,
but that's outside the scope of this PR), so restore the old _Int128
type and continue using it as an implementation detail for now.
We can switch over sometime post-6.0 release.

Also underscores the _low and _high properties and the corresponding
initializer, as @lorentey correctly requested.

stephentyrone commented:
@swift-ci test

stephentyrone commented:
We're going to do C importing in separate PRs (since the C ABI for _BitInt(128) is still in flux), so we don't need `CVarArg...` yet.

stephentyrone committed:
Add CustomReflectable conformance for [U]Int128

stephentyrone commented:
@swift-ci test

stephentyrone commented:
On wasm32 targets, yes, it will use the pair storage representation.
